File: solutions/go/MedianOfTwoSortedArrays.go
func findMedianSortedArrays(x, y []int) float64 {
	var median float64

	lenx, leny := len(x), len(y)
	if lenx > leny {
		return findMedianSortedArrays(y, x)
	}

	low, high := 0, lenx

	for low <= high {
		pivotx := (low + high) / 2
		pivoty := (lenx + leny + 1) / 2 - pivotx

		leftx, rightx := math.MinInt32, math.MaxInt32
		lefty, righty := math.MinInt32, math.MaxInt32

		if pivotx != 0 {
			leftx = x[pivotx-1]
		}

		if pivotx != lenx {
			rightx = x[pivotx]
		}

		if pivoty != 0 {
			lefty = y[pivoty-1]
		}

		if pivoty != leny {
			righty = y[pivoty]
		}

		if leftx <= righty && lefty <= rightx {
			if (lenx + leny) & 1 == 1 {
				median = float64(max(leftx, lefty))
			} else {
				median = float64(max(leftx, lefty) + min(rightx, righty)) / 2
			}

			break
		} else if leftx > righty {
			high = pivotx - 1
		} else {
			low = pivotx + 1
		}
	}

	return median
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}

	return b
}

File: solutions/c/MedianOfTwoSortedArrays.c
#define max(a, b) ((a) > (b) ? (a) : (b))
#define min(a, b) ((a) < (b) ? (a) : (b))

double
findMedianSortedArrays(int *x, int lenx, int *y, int leny)
{
    if (lenx > leny) {
        return findMedianSortedArrays(y, leny, x, lenx);
    }

    int low, high;
    double median;

    low = 0;
    high = lenx;

    while (low <= high) {
        int pivotx = (low + high) / 2;
        int pivoty = (lenx + leny + 1) / 2 - pivotx;

        int leftx = pivotx == 0 ? INT_MIN : x[pivotx-1];
        int rightx = pivotx == lenx ? INT_MAX : x[pivotx];
        int lefty = pivoty == 0 ? INT_MIN : y[pivoty-1];
        int righty = pivoty == leny ? INT_MAX : y[pivoty];

        if (leftx <= righty && lefty <= rightx) {
            if (((lenx+leny) & 1) != 0) {
                median = (double)max(leftx, lefty);
            } else {
                median = (double)(max(leftx, lefty) + min(rightx, righty)) / 2;
            }

            break;
        } else if (leftx > righty) {
            high = pivotx - 1;
        } else {
            low = pivotx + 1;
        }
    }

    return median;
}

